<a href="https://colab.research.google.com/github/arsalan-jamal002/SMIT-AI-DATA-SCIENCE/blob/main/Loan.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
#Loading and importing data
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os

In [3]:
# Check available files
files = [f for f in os.listdir('.')
 if os.path.isfile(f)]
print("Available files:", files)

Available files: ['loan-train.csv']


In [4]:
# Load dataset
df = pd.read_csv('loan-train.csv')


In [5]:
#print datashape
df.shape

(1614, 13)

In [9]:
#columns list
df.columns.tolist()

['Loan_ID',
 'Gender',
 'Married',
 'Dependents',
 'Education',
 'Self_Employed',
 'ApplicantIncome',
 'CoapplicantIncome',
 'LoanAmount',
 'Loan_Amount_Term',
 'Credit_History',
 'Property_Area',
 'Loan_Status']

In [6]:
#print first 5 rows of data
df.head(5)

,Loan_ID,Gender,Married,Dependents,Education,Self_Employed,ApplicantIncome,CoapplicantIncome,LoanAmount,Loan_Amount_Term,Credit_History,Property_Area,Loan_Status
0,LP001002,Male,No,0,Graduate,No,5849,0.0,NaN,360.0,1.0,Urban,Y
1,LP001003,Male,Yes,1,Graduate,No,4583,1508.0,128.0,360.0,1.0,Rural,N
2,LP001005,Male,Yes,0,Graduate,Yes,3000,0.0,66.0,360.0,1.0,Urban,Y
3,LP001006,Male,Yes,0,Not Graduate,No,2583,2358.0,120.0,360.0,1.0,Urban,Y
4,LP001008,Male,No,0,Graduate,No,6000,0.0,141.0,360.0,1.0,Urban,Y


In [7]:
#data info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1614 entries, 0 to 1613
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Loan_ID            1614 non-null   object 
 1   Gender             1601 non-null   object 
 2   Married            1611 non-null   object 
 3   Dependents         1599 non-null   object 
 4   Education          1614 non-null   object 
 5   Self_Employed      1582 non-null   object 
 6   ApplicantIncome    1614 non-null   int64  
 7   CoapplicantIncome  1614 non-null   float64
 8   LoanAmount         1592 non-null   float64
 9   Loan_Amount_Term   1600 non-null   float64
 10  Credit_History     1564 non-null   float64
 11  Property_Area      1614 non-null   object 
 12  Loan_Status        1614 non-null   object 
dtypes: float64(4), int64(1), object(8)
memory usage: 164.1+ KB


In [8]:
#sum of missing values
df.isnull().sum()

,0
Loan_ID,0
Gender,13
Married,3
Dependents,15
Education,0
Self_Employed,32
ApplicantIncome,0
CoapplicantIncome,0
LoanAmount,22
Loan_Amount_Term,14


In [12]:
print("\nDuplicate rows:", df.duplicated().sum())
print("Duplicate Loan_IDs:", df['Loan_ID'].duplicated().sum())


Duplicate rows: 0
Duplicate Loan_IDs: 76


In [11]:
print("Duplicate Loan_IDs:", df['Loan_ID'].duplicated().sum())

Duplicate Loan_IDs: 76


In [13]:
print("\nValue counts for categorical columns:")
cat_cols = ['Gender', 'Married', 'Dependents', 'Education', 'Self_Employed', 'Credit_History', 'Property_Area', 'Loan_Status']
for col in cat_cols:
    print(f"\n--- {col} ---")
    print(df[col].value_counts(dropna=False))


Value counts for categorical columns:

--- Gender ---
Gender
Male      980
Female    621
NaN        13
Name: count, dtype: int64

--- Married ---
Married
Yes    925
No     686
NaN      3
Name: count, dtype: int64

--- Dependents ---
Dependents
0      619
1      341
2      336
3+     303
NaN     15
Name: count, dtype: int64

--- Education ---
Education
Graduate        952
Not Graduate    662
Name: count, dtype: int64

--- Self_Employed ---
Self_Employed
No     999
Yes    583
NaN     32
Name: count, dtype: int64

--- Credit_History ---
Credit_History
1.0    994
0.0    570
NaN     50
Name: count, dtype: int64

--- Property_Area ---
Property_Area
Urban        549
Semiurban    545
Rural        520
Name: count, dtype: int64

--- Loan_Status ---
Loan_Status
Y    915
N    699
Name: count, dtype: int64


In [14]:
print("\nSummary statistics for numerical columns:")
print(df.describe())


Summary statistics for numerical columns:
       ApplicantIncome  CoapplicantIncome   LoanAmount  Loan_Amount_Term  \
count      1614.000000        1614.000000  1592.000000       1600.000000   
mean       8746.304833        3658.722999   218.846734        290.475000   
std        6249.885122        3291.745935   133.623293         81.899735   
min         150.000000           0.000000     9.000000         12.000000   
25%        3675.250000         986.250000   112.000000        240.000000   
50%        7045.000000        2941.500000   180.000000        360.000000   
75%       13251.000000        5893.000000   324.000000        360.000000   
max       81000.000000       41667.000000   700.000000        480.000000   

       Credit_History  
count      1564.00000  
mean          0.63555  
std           0.48143  
min           0.00000  
25%           0.00000  
50%           1.00000  
75%           1.00000  
max           1.00000  


In [15]:
# Let's explore why there are duplicate Loan_IDs
duplicates = df[df.duplicated('Loan_ID', keep=False)].sort_values('Loan_ID')
print(duplicates.head(10))

      Loan_ID  Gender Married Dependents     Education Self_Employed  \
614  LP003002    Male     Yes          0  Not Graduate           Yes   
615  LP003002  Female     Yes          2      Graduate            No   
618  LP003029  Female     Yes          0      Graduate           Yes   
619  LP003029    Male     Yes          0      Graduate           Yes   
631  LP003103    Male      No          2  Not Graduate            No   
632  LP003103  Female      No          2  Not Graduate           Yes   
638  LP003139    Male      No          2  Not Graduate           Yes   
639  LP003139  Female      No         3+      Graduate            No   
666  LP003339    Male      No          1  Not Graduate            No   
665  LP003339  Female      No          2  Not Graduate            No   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
614             8295             6148.0       330.0             360.0   
615            13525             6635.0       217.0          

In [16]:
# Let's explore why there are duplicate Loan_IDs
duplicates = df[df.duplicated('Loan_ID', keep=False)].sort_values('Loan_ID')
print(duplicates.head(10))

      Loan_ID  Gender Married Dependents     Education Self_Employed  \
614  LP003002    Male     Yes          0  Not Graduate           Yes   
615  LP003002  Female     Yes          2      Graduate            No   
618  LP003029  Female     Yes          0      Graduate           Yes   
619  LP003029    Male     Yes          0      Graduate           Yes   
631  LP003103    Male      No          2  Not Graduate            No   
632  LP003103  Female      No          2  Not Graduate           Yes   
638  LP003139    Male      No          2  Not Graduate           Yes   
639  LP003139  Female      No         3+      Graduate            No   
666  LP003339    Male      No          1  Not Graduate            No   
665  LP003339  Female      No          2  Not Graduate            No   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
614             8295             6148.0       330.0             360.0   
615            13525             6635.0       217.0          

In [18]:
# Check distributions and outliers using clean visualization code
sns.set_theme(style='whitegrid')

In [22]:
# Let's save a couple of plots representing the key insights
# 1. Distribution of Numerical Features
fig, axes = plt.subplots(2, 2, figsize=(14, 10))
sns.histplot(df['ApplicantIncome'], kde=True, ax=axes[0,0], color='blue').set_title('Applicant Income Distribution')
sns.histplot(df['CoapplicantIncome'], kde=True, ax=axes[0,1], color='orange').set_title('Coapplicant Income Distribution')
sns.histplot(df['LoanAmount'].dropna(), kde=True, ax=axes[1,0], color='green').set_title('Loan Amount Distribution')
sns.boxplot(x='Loan_Status', y='ApplicantIncome', data=df, ax=axes[1,1]).set_title('ApplicantIncome vs Loan Status')
plt.tight_layout()
fig.savefig('numerical_distributions.png')
plt.close()

In [23]:
# 2. Categorical features count plots
fig, axes = plt.subplots(2, 3, figsize=(16, 10))
sns.countplot(x='Gender', hue='Loan_Status', data=df, ax=axes[0,0])
sns.countplot(x='Married', hue='Loan_Status', data=df, ax=axes[0,1])
sns.countplot(x='Education', hue='Loan_Status', data=df, ax=axes[0,2])
sns.countplot(x='Self_Employed', hue='Loan_Status', data=df, ax=axes[1,0])
sns.countplot(x='Credit_History', hue='Loan_Status', data=df, ax=axes[1,1])
sns.countplot(x='Property_Area', hue='Loan_Status', data=df, ax=axes[1,2])
plt.tight_layout()
fig.savefig('categorical_distributions.png')
plt.close()
print("Plots saved successfully.")

Plots saved successfully.


In [24]:
# Missing values handling scheme
df['Gender'] = df['Gender'].fillna(df['Gender'].mode()[0])
df['Married'] = df['Married'].fillna(df['Married'].mode()[0])
df['Dependents'] = df['Dependents'].fillna(df['Dependents'].mode()[0])
df['Self_Employed'] = df['Self_Employed'].fillna(df['Self_Employed'].mode()[0])
df['Credit_History'] = df['Credit_History'].fillna(df['Credit_History'].mode()[0])

df['LoanAmount'] = df['LoanAmount'].fillna(df['LoanAmount'].median())
df['Loan_Amount_Term'] = df['Loan_Amount_Term'].fillna(df['Loan_Amount_Term'].median())

# 1. **Dataset Overview**
Total Rows (Records): 1,614

Total Columns: 13

Duplicate Rows: No exact duplicate rows were found, but 76 Loan_IDs are repeated with different structural details (these duplicate entries can be cleaned later if required).

2. Column-by-Column EDA (Detailed Analysis)
Categorical Columns Analysis:
Loan_ID: This is the unique ID for each applicant. It contains 76 repeated keys where different profiles exist.

Gender: This column contains 980 Male and 621 Female records, with 13 missing values.

Married: 925 applicants are married, while 686 are unmarried. There are 3 missing values.

Dependents: Indicates the number of dependents. The distribution is: 0 (619), 1 (341), 2 (336), and 3+ (303) records. There are 15 missing values.

Education: 952 applicants are Graduates and 662 are Not Graduates. There are no missing values.

Self_Employed: 999 applicants are job holders, while 583 run their own business (self-employed). There are 32 missing values.

Credit_History: This is the most critical feature for loan approval. 994 applicants have a good credit history (1.0), while 570 have a poor credit history (0.0). There are 50 missing values.

Property_Area: Semiurban (545), Urban (549), and Rural (520) areas are almost equally distributed. There are no missing values.

Loan_Status (Target Variable): 915 loans were approved (Y) and 699 were rejected (N).

Numerical Columns Analysis:
ApplicantIncome: The minimum income is 150 and the maximum is 81,000. The median income is 7,045, which indicates that the data is right-skewed and contains some high-earning outliers.

CoapplicantIncome: The minimum is 0 and the maximum is 41,667. In many cases, the coapplicant income is 0 (Median: 2,941.5).

LoanAmount: The minimum is 9 and the maximum is 700 (in thousands). There are 22 missing values.

Loan_Amount_Term: Most of the loans are for 360 months (30 years). There are 14 missing values.

3. Visualizations
Two main plots have been generated to understand the distributions and patterns in your data:

numerical_distributions.png: Histograms/boxplots have been created here to visualize the skewness and outliers for Applicant Income, Coapplicant Income, and Loan Amount.

categorical_distributions.png: This shows the relationship between Gender, Marriage status, Education, and Credit History with the Loan Status (e.g., applicants with a Credit History of 1.0 have a very high loan approval rate).

4. Data Cleaning Step
To handle the missing values and complete the analysis, we can fill the numerical values with their median and the categorical values with their mode (the most frequent value).